# Assignment 4

In [1]:
import networkx as nx
import pandas as pd
import numpy as np
import pickle

---

## Part 1 - Random Graph Identification

For the first part of this assignment you will analyze randomly generated graphs and determine which algorithm created them.

In [2]:
G1 = nx.read_gpickle("assets/A4_P1_G1")
G2 = nx.read_gpickle("assets/A4_P1_G2")
G3 = nx.read_gpickle("assets/A4_P1_G3")
G4 = nx.read_gpickle("assets/A4_P1_G4")
G5 = nx.read_gpickle("assets/A4_P1_G5")
P1_Graphs = [G1, G2, G3, G4, G5]

<br>
`P1_Graphs` is a list containing 5 networkx graphs. Each of these graphs were generated by one of three possible algorithms:
* Preferential Attachment (`'PA'`)
* Small World with low probability of rewiring (`'SW_L'`)
* Small World with high probability of rewiring (`'SW_H'`)

Anaylze each of the 5 graphs using any methodology and determine which of the three algorithms generated each graph.

*The `graph_identification` function should return a list of length 5 where each element in the list is either `'PA'`, `'SW_L'`, or `'SW_H'`.*

In [3]:
import re

[i for i in nx.__dir__() if re.search('layout$', i)]

['layout',
 'bipartite_layout',
 'circular_layout',
 'kamada_kawai_layout',
 'random_layout',
 'rescale_layout',
 'shell_layout',
 'spring_layout',
 'spectral_layout',
 'planar_layout',
 'fruchterman_reingold_layout',
 'spiral_layout',
 'multipartite_layout']

In [4]:
#PA: degree high & clust low
#SW_L: degree mid & clust mid
#SW_H: degree high & clust high

In [5]:
import numpy as np

def graph_identification():
    
    G1_avg_degree = sum(dict(nx.degree(G1)).values())/len(nx.nodes(G1)) #sum(nx.average_degree_connectivity(G1).values())/len(nx.average_degree_connectivity(G1))
    G1_avg_clust = nx.average_clustering(G1)
    
    G2_avg_degree = sum(dict(nx.degree(G2)).values())/len(nx.nodes(G2)) #sum(nx.average_degree_connectivity(G1).values())/len(nx.average_degree_connectivity(G1))
    G2_avg_clust = nx.average_clustering(G2)

    G3_avg_degree = sum(dict(nx.degree(G3)).values())/len(nx.nodes(G3)) #sum(nx.average_degree_connectivity(G1).values())/len(nx.average_degree_connectivity(G1))
    G3_avg_clust = nx.average_clustering(G3)

    G4_avg_degree = sum(dict(nx.degree(G4)).values())/len(nx.nodes(G4)) #sum(nx.average_degree_connectivity(G1).values())/len(nx.average_degree_connectivity(G1))
    G4_avg_clust = nx.average_clustering(G4)

    G5_avg_degree = sum(dict(nx.degree(G5)).values())/len(nx.nodes(G5)) #sum(nx.average_degree_connectivity(G1).values())/len(nx.average_degree_connectivity(G1))
    G5_avg_clust = nx.average_clustering(G5)
    
    #return ({'G1': [G1_avg_degree, np.round(G1_avg_clust,4)], 'G2': [G2_avg_degree, np.round(G2_avg_clust,4)], 
            #'G3': [G3_avg_degree, np.round(G3_avg_clust,4)], 'G4': [G4_avg_degree, np.round(G4_avg_clust,4)], 'G5': [G5_avg_degree, np.round(G5_avg_clust,4)]})


    return ['PA', 'SW_L', 'SW_L', 'PA', 'SW_H'] #nx.draw_networkx(G1, pos=nx.circular_layout(G1), node_size=0.1, width=0.05) 

graph_identification()

['PA', 'SW_L', 'SW_L', 'PA', 'SW_H']

In [6]:
def t():
    methods = []
    for G in P1_Graphs:
        degrees = G.degree()
        degree_values = sorted(set(dict(degrees).values()))
        degree_hist = [list(dict(degrees).values()).count(i) / float(nx.number_of_nodes(G)) for i in degree_values]
        clustering = nx.average_clustering(G)
        shortest_path = nx.average_shortest_path_length(G)
        print(G, degree_values, degree_hist, clustering, shortest_path)
        
        if len(degree_hist)>10:
            methods.append('PA')
        elif clustering < 0.1:
            methods.append('SW_H')
        else:
            methods.append('SW_L')
    return methods
t()

Graph with 1000 nodes and 999 edges [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 13, 14, 15, 20, 21, 25, 30, 35, 46, 54] [0.684, 0.151, 0.067, 0.031, 0.013, 0.017, 0.006, 0.008, 0.004, 0.002, 0.004, 0.003, 0.001, 0.002, 0.001, 0.001, 0.001, 0.001, 0.001, 0.001, 0.001] 0.0 6.530506506506507
Graph with 1000 nodes and 2000 edges [2, 3, 4, 5] [0.001, 0.01, 0.977, 0.012] 0.49310000000000004 43.80284684684685
Graph with 1000 nodes and 2000 edges [3, 4, 5] [0.013, 0.974, 0.013] 0.4897333333333334 39.007695695695695
Graph with 1000 nodes and 999 edges [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 16, 17, 19, 22, 23, 24, 25, 30, 36, 39] [0.661, 0.166, 0.073, 0.033, 0.017, 0.013, 0.011, 0.002, 0.004, 0.001, 0.007, 0.002, 0.001, 0.001, 0.001, 0.001, 0.001, 0.001, 0.001, 0.001, 0.001, 0.001] 0.0 8.158990990990992
Graph with 1000 nodes and 2000 edges [2, 3, 4, 5, 6, 7] [0.01, 0.154, 0.68, 0.139, 0.016, 0.001] 0.36504285714285717 8.532046046046046


['PA', 'SW_L', 'SW_L', 'PA', 'SW_L']

In [7]:
ans_one = graph_identification()
assert type(ans_one) == list, "You must return a list"


---

## Part 2 - Company Emails

For the second part of this assignment you will be working with a company's email network where each node corresponds to a person at the company, and each edge indicates that at least one email has been sent between two people.

The network also contains the node attributes `Department` and `ManagmentSalary`.

`Department` indicates the department in the company which the person belongs to, and `ManagmentSalary` indicates whether that person is receiving a managment position salary.

In [8]:
G = pickle.load(open('assets/email_prediction_NEW.txt', 'rb'))

print(f"Graph with {len(nx.nodes(G))} nodes and {len(nx.edges(G))} edges")

Graph with 1005 nodes and 16706 edges


In [9]:
print(nx.info(G))

Graph with 1005 nodes and 16706 edges


### Part 2A - Salary Prediction

Using network `G`, identify the people in the network with missing values for the node attribute `ManagementSalary` and predict whether or not these individuals are receiving a managment position salary.

To accomplish this, you will need to create a matrix of node features of your choice using networkx, train a sklearn classifier on nodes that have `ManagementSalary` data, and predict a probability of the node receiving a managment salary for nodes where `ManagementSalary` is missing.



Your predictions will need to be given as the probability that the corresponding employee is receiving a managment position salary.

The evaluation metric for this assignment is the Area Under the ROC Curve (AUC).

Your grade will be based on the AUC score computed for your classifier. A model which with an AUC of 0.75 or higher will recieve full points.

Using your trained classifier, return a Pandas series of length 252 with the data being the probability of receiving managment salary, and the index being the node id.

    Example:
    
        1       1.0
        2       0.0
        5       0.8
        8       1.0
            ...
        996     0.7
        1000    0.5
        1001    0.0
        Length: 252, dtype: float64

In [10]:
list(G.nodes(data=True))[:5] # print the first 5 nodes

[(0, {'Department': 1, 'ManagementSalary': 0.0}),
 (1, {'Department': 1, 'ManagementSalary': nan}),
 (581, {'Department': 3, 'ManagementSalary': 0.0}),
 (6, {'Department': 25, 'ManagementSalary': 1.0}),
 (65, {'Department': 4, 'ManagementSalary': nan})]

In [11]:
def salary_predictions():
    from sklearn.preprocessing import StandardScaler
    from sklearn.ensemble import RandomForestClassifier
    from sklearn.model_selection import train_test_split
    from sklearn.linear_model import LinearRegression, LogisticRegression
    from sklearn.svm import SVC
    from sklearn.dummy import DummyClassifier
    from sklearn.metrics import precision_recall_curve, roc_curve, roc_auc_score

    df1 = pd.DataFrame(G.nodes(data=True))
    df1['Dep'] = df1[1].apply(lambda x: x['Department'])
    df1['Management'] = df1[1].apply(lambda x: x['ManagementSalary'])
    df1.drop(columns=[1], inplace=True)
    df1.rename(columns={0:'Node'}, inplace=True)
    
    test_data = df1[df1['Management'].isna()].drop(columns=('Management'))
    
    training_data = df1[~df1['Management'].isna()]
    training_data_X = training_data.drop(columns='Management')
    training_data_y = training_data['Management']

    
    X_train, X_test, y_train, y_test = train_test_split(training_data_X, training_data_y, random_state=0)
    
    
    #svc = SVC(C=1, gamma=10).fit(X_train, y_train)
    #dummy = DummyClassifier(strategy='most_frequent').fit(X_train, y_train)

    logreg = LogisticRegression().fit(X_train, y_train)
    
    pred = logreg.predict_proba(test_data)

    resultdf = test_data.set_index('Node').drop(columns=('Dep'))
    resultdf['pred'] = pred[:,1] #just consider the second column, the prob of the result being 1 (positive)
    
    #roc_auc = roc_auc_score(pred, y_test)  # Provide true labels (y_true) for test data
    #pred = logreg.predict(X_test)

    return resultdf.squeeze() #logreg.score(X_test, y_test)

salary_predictions()

Node
1      0.285203
65     0.269042
18     0.275063
215    0.174821
283    0.380914
         ...   
691    0.096588
788    0.082109
944    0.051577
798    0.036389
808    0.067987
Name: pred, Length: 252, dtype: float64

In [12]:
ans_salary_preds = salary_predictions()
assert type(ans_salary_preds) == pd.core.series.Series, "You must return a Pandas series"
assert len(ans_salary_preds) == 252, "The series must be of length 252"


### Part 2B - New Connections Prediction

For the last part of this assignment, you will predict future connections between employees of the network. The future connections information has been loaded into the variable `future_connections`. The index is a tuple indicating a pair of nodes that currently do not have a connection, and the `Future Connection` column indicates if an edge between those two nodes will exist in the future, where a value of 1.0 indicates a future connection.

In [13]:
future_connections = pd.read_csv('assets/Future_Connections.csv', index_col=0, converters={0: eval})
future_connections.head(10)

Future Connection
(6, 840)                  0.0
(4, 197)                  0.0
(620, 979)                0.0
(519, 872)                0.0
(382, 423)                0.0
(97, 226)                 1.0
(349, 905)                0.0
(429, 860)                0.0
(309, 989)                0.0
(468, 880)                0.0

In [14]:
len(future_connections)

488446

Using network `G` and `future_connections`, identify the edges in `future_connections` with missing values and predict whether or not these edges will have a future connection.

To accomplish this, you will need to:      
1. Create a matrix of features of your choice for the edges found in `future_connections` using Networkx     
2. Train a sklearn classifier on those edges in `future_connections` that have `Future Connection` data     
3. Predict a probability of the edge being a future connection for those edges in `future_connections` where `Future Connection` is missing.



Your predictions will need to be given as the probability of the corresponding edge being a future connection.

The evaluation metric for this assignment is the Area Under the ROC Curve (AUC).

Your grade will be based on the AUC score computed for your classifier. A model which with an AUC of 0.75 or higher will recieve full points.

Using your trained classifier, return a series of length 122112 with the data being the probability of the edge being a future connection, and the index being the edge as represented by a tuple of nodes.

    Example:
    
        (107, 348)    0.35
        (542, 751)    0.40
        (20, 426)     0.55
        (50, 989)     0.35
                  ...
        (939, 940)    0.15
        (555, 905)    0.35
        (75, 101)     0.65
        Length: 122112, dtype: float64

In [15]:
future_connections

Future Connection
(6, 840)                  0.0
(4, 197)                  0.0
(620, 979)                0.0
(519, 872)                0.0
(382, 423)                0.0
...                       ...
(165, 923)                NaN
(673, 755)                NaN
(939, 940)                NaN
(555, 905)                NaN
(75, 101)                 NaN

[488446 rows x 1 columns]

In [16]:
def new_connections_predictions():
    from sklearn.preprocessing import StandardScaler, LabelEncoder
    from sklearn.ensemble import RandomForestClassifier
    from operator import itemgetter
    from sklearn.model_selection import train_test_split
    from sklearn.linear_model import LinearRegression, LogisticRegression
    from sklearn.svm import SVC
    from sklearn.dummy import DummyClassifier
    from sklearn.metrics import precision_recall_curve, roc_curve, roc_auc_score
    
    
    df2 = future_connections.copy()
    
    df2['send'] = list(map(itemgetter(0),df2.index))
    df2['receive'] = list(map(itemgetter(1),df2.index))
    df2['fut_connection'] = df2['Future Connection']
    df2.drop(columns=('Future Connection'), inplace=True)
    
    t = pd.DataFrame(nx.degree(G)).rename(columns={0:'node', 1:'degree'})
    t['closeness_c'] = nx.closeness_centrality(G).values()
    t['betweenness_c'] = nx.betweenness_centrality(G).values()
    t['degree_c'] = nx.degree_centrality(G).values()

    #nx.shortest_path_length(G, source=6, target=840)
    
    test = df2.merge(t, left_on='send', right_on='node', how='left')
    test.rename(columns={'degree': 's_degree', 'closeness_c':'s_closeness_c', 'betweenness_c':'s_betweenness_c', 'degree_c':'s_degree_c'}, inplace=True)
    test2 = test.merge(t, left_on='receive', right_on='node', how='left')
    
    finaldf = test2.drop(columns=['node_x', 'node_y'])
    
    trainingdf = finaldf[~finaldf['fut_connection'].isna()]
    testingdf = finaldf[finaldf['fut_connection'].isna()].drop(columns=('fut_connection'))
    
    trainingdf_X = trainingdf[['send', 'receive', 's_degree', 's_closeness_c', 's_betweenness_c', 's_degree_c', 'degree', 'closeness_c', 'betweenness_c', 'degree_c']]
    trainingdf_y = trainingdf['fut_connection']
    
    X_train, X_test, y_train, y_test = train_test_split(trainingdf_X, trainingdf_y, random_state=0)
    
    logreg = RandomForestClassifier().fit(X_train, y_train)
    score = logreg.score(X_test, y_test)
    
    pred = logreg.predict_proba(testingdf)[:,1]
    
    testingdf['prediction'] = pred
    testingdf['new_index'] = testingdf.apply(lambda row: (round(row['send']), round(row['receive'])), axis=1)
        
    return testingdf.set_index('new_index')['prediction'].squeeze()

new_connections_predictions()

new_index
(107, 348)    0.31
(542, 751)    0.16
(20, 426)     0.53
(50, 989)     0.00
(942, 986)    0.01
              ... 
(165, 923)    0.10
(673, 755)    0.00
(939, 940)    0.00
(555, 905)    0.00
(75, 101)     0.01
Name: prediction, Length: 122112, dtype: float64

In [ ]:
ans_prob_preds = new_connections_predictions()
assert type(ans_prob_preds) == pd.core.series.Series, "You must return a Pandas series"
assert len(ans_prob_preds) == 122112, "The series must be of length 122112"
